jumanji/environments/packing/jigsaw/env.py

```diff
@@ -0,0 +1,523 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Sequence, Tuple
+
+import chex
+import jax
+import jax.numpy as jnp
+import matplotlib
+from numpy.typing import NDArray
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.packing.jigsaw.generator import (
+    InstanceGenerator,
+    RandomJigsawGenerator,
+)
+from jumanji.environments.packing.jigsaw.reward import DenseReward, RewardFn
+from jumanji.environments.packing.jigsaw.types import Observation, State
+from jumanji.environments.packing.jigsaw.utils import compute_grid_dim, rotate_piece
+from jumanji.environments.packing.jigsaw.viewer import JigsawViewer
+from jumanji.types import TimeStep, restart, termination, transition
+from jumanji.viewer import Viewer
+
+
+class Jigsaw(Environment[State]):
+
+    """A Jigsaw solving environment with a configurable number of row and column pieces.
+    Here the goal of an agent is to solve a jigsaw puzzle by placing pieces
```

Comment:
@clement-bonnet asked that we just tab lines following the first line.

Reply:
Done :)